server/graphql/utils/request.ts

```diff
@@ -20,19 +20,69 @@ const getRequestsById = (requestIds, dataSources) => {
   return requestIds.map(id => dataSources.requests.getById(id))
 }
 
-const updateRequestHelper = (request, dataSources): Promise<ServerResponseInterface> => {
+const revertUpdates = (request, dataSources, currentRequestCopy, oldRequestTypeCopy, newRequestTypeCopy) => {
+  dataSources.requests.update(currentRequestCopy)
+  if(request.requestType) {
+    dataSources.requestTypes.update(oldRequestTypeCopy)
+    dataSources.requestTypes.update(newRequestTypeCopy)
+  }
+}
+
+const updateRequestHelper = async (request, dataSources): Promise<Document> => {
   if(!request.id) {
     throw new UserInputError('Missing argument value', { argumentName: 'id' })
   }
-  const requestTypeId = dataSources.requests.getById(request.id.toString()).requestType
-  return dataSources.requests.update(request)
-    .then(res => {
-      updateRequestTypeHelper({"id": requestTypeId}, dataSources)
-      return res
-    })
-    .catch(error => {
-      return error
-    })
+
+  const currentRequest = dataSources.requests.getById(request.id.toString())
+  const currentRequestCopy = currentRequest.toObject()
+  currentRequestCopy.id = currentRequestCopy._id
+  let oldRequestTypeCopy, newRequestTypeCopy
+
+  if(request.requestType) {
+    // If these fail then nothing has been changed at this point, so nothing to revert
+    oldRequestTypeCopy = dataSources.requestTypes.getById(currentRequest.requestType.toString()).toObject()
+    oldRequestTypeCopy.id = oldRequestTypeCopy._id
+    newRequestTypeCopy = dataSources.requestTypes.getById(request.requestType.toString()).toObject()
+    newRequestTypeCopy.id = newRequestTypeCopy._id
+  }
+
+  if(request.fulfilled === true && currentRequest.fulfilled === false) {
+    request.dateFulfilled = Date.now()
+  }
+
+  try {
+
+    if(request.requestType) {
+      const oldRequestType = dataSources.requestTypes.getById(currentRequest.requestType.toString())
+      const newRequestType = dataSources.requestTypes.getById(request.requestType.toString())
+      oldRequestType.requests = oldRequestType.requests.filter(id => !id.equals(request.id))
+      newRequestType.requests.push(request.id)
+      await dataSources.requestTypes.update(oldRequestType)
+      await dataSources.requestTypes.update(newRequestType)
```

Comment:
I put await here because sometimes the rollback update finishes before this update, which causes the wrong version to be in the database.